core-text/src/run.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2013 The Servo Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::borrow::Cow;
+use std::os::raw::c_void;
+use std::slice;
+use std::ptr;
+use core_foundation::base::{CFIndex, CFTypeID, TCFType, CFType, CFRange};
+use core_foundation::dictionary::{CFDictionary, CFDictionaryRef};
+use core_foundation::string::CFString;
+use core_graphics::font::CGGlyph;
+use core_graphics::geometry::CGPoint;
+
+#[repr(C)]
+pub struct __CTRun(c_void);
+
+pub type CTRunRef = *const __CTRun;
+
+declare_TCFType! {
+    CTRun, CTRunRef
+}
+impl_TCFType!(CTRun, CTRunRef, CTRunGetTypeID);
+impl_CFTypeDescription!(CTRun);
+
+impl CTRun {
+    pub fn attributes(&self) -> Option<CFDictionary<CFString, CFType>> {
+        unsafe {
+            let attrs = CTRunGetAttributes(self.0);
+            if attrs.is_null() {
+                return None;
+            }
+            Some(TCFType::wrap_under_get_rule(attrs))
+        }
+    }
+    pub fn glyph_count(&self) -> CFIndex {
+        unsafe {
+            CTRunGetGlyphCount(self.0)
+        }
+    }
+
+    pub fn glyphs(&self) -> Cow<[CGGlyph]> {
+        unsafe {
+            // CTRunGetGlyphsPtr can return null under some not understood circumstances.
+            // If it does the Apple documentation tells us to allocate our own buffer and call
+            // CTRunGetGlyphs
+            let count = CTRunGetGlyphCount(self.0);
+            let glyphs_ptr = CTRunGetGlyphsPtr(self.0);
+            if glyphs_ptr != ptr::null() {
+                Cow::from(slice::from_raw_parts(glyphs_ptr, count as usize))
+            } else {
+                let mut vec = Vec::with_capacity(count as usize);
+                // "If the length of the range is set to 0, then the copy operation will continue
+                // from the start index of the range to the end of the run"
+                CTRunGetGlyphs(self.0, CFRange::init(0, 0), vec.as_mut_ptr());
```

Comment:
We need to call `vec.set_len(count)` to have a usable vector after this operation.